spring-ai-core/src/main/java/org/springframework/ai/tool/function/FunctionToolCallback.java

```diff
@@ -0,0 +1,212 @@
+
+	private final ToolMetadata toolMetadata;
+
+	private final Type toolInputType;
+
+	private final BiFunction<I, ToolContext, O> toolFunction;
+
+	private final ToolCallResultConverter toolCallResultConverter;
+
+	public FunctionToolCallback(ToolDefinition toolDefinition, @Nullable ToolMetadata toolMetadata, Type toolInputType,
+			BiFunction<I, ToolContext, O> toolFunction, @Nullable ToolCallResultConverter toolCallResultConverter) {
+		Assert.notNull(toolDefinition, "toolDefinition cannot be null");
+		Assert.notNull(toolInputType, "toolInputType cannot be null");
+		Assert.notNull(toolFunction, "toolFunction cannot be null");
+		this.toolDefinition = toolDefinition;
+		this.toolMetadata = toolMetadata != null ? toolMetadata : DEFAULT_TOOL_METADATA;
+		this.toolFunction = toolFunction;
+		this.toolInputType = toolInputType;
+		this.toolCallResultConverter = toolCallResultConverter != null ? toolCallResultConverter
+				: DEFAULT_RESULT_CONVERTER;
+	}
+
+	@Override
+	public ToolDefinition getToolDefinition() {
+		return toolDefinition;
+	}
+
+	@Override
+	public ToolMetadata getToolMetadata() {
+		return toolMetadata;
+	}
+
+	@Override
+	public String call(String toolInput) {
+		return call(toolInput, null);
+	}
+
+	@Override
+	public String call(String toolInput, @Nullable ToolContext toolContext) {
+		Assert.hasText(toolInput, "toolInput cannot be null or empty");
+
+		logger.debug("Starting execution of tool: {}", toolDefinition.name());
+
+		I request = JsonParser.fromJson(toolInput, toolInputType);
+		O response = toolFunction.apply(request, toolContext);
+
+		logger.debug("Successful execution of tool: {}", toolDefinition.name());
+
+		return toolCallResultConverter.apply(response, null);
+	}
+
+	/**
+	 * Build a {@link FunctionToolCallback} from a {@link BiFunction}.
+	 */
+	public static <I, O> Builder<I, O> builder(String name, BiFunction<I, ToolContext, O> function) {
+		return new Builder<>(name, function);
+	}
+
+	/**
+	 * Build a {@link FunctionToolCallback} from a {@link Function}.
+	 */
+	public static <I, O> Builder<I, O> builder(String name, Function<I, O> function) {
+		Assert.notNull(function, "function cannot be null");
+		return new Builder<>(name, (request, context) -> function.apply(request));
+	}
+
+	/**
+	 * Build a {@link FunctionToolCallback} from a {@link Supplier}.
+	 */
+	public static <O> Builder<Void, O> builder(String name, Supplier<O> supplier) {
+		Assert.notNull(supplier, "supplier cannot be null");
+		Function<Void, O> function = input -> supplier.get();
+		return builder(name, function).inputType(Void.class);
+	}
+
+	/**
+	 * Build a {@link FunctionToolCallback} from a {@link Consumer}.
+	 */
+	public static <I> Builder<I, Void> builder(String name, Consumer<I> consumer) {
+		Assert.notNull(consumer, "consumer cannot be null");
+		Function<I, Void> function = (I input) -> {
+			consumer.accept(input);
+			return null;
+		};
+		return builder(name, function);
+	}
+
+	public static class Builder<I, O> {
```

Comment:
I kept a similar API to the one in FunctionCallback.Builder so to make it straightforward to migrate to this new Builder.